src/shared/tests.c

```diff
@@ -41,6 +42,16 @@ const char* get_exe_relative_testdata_dir(void) {
         static char testdir[PATH_MAX];
 
         assert_se(readlink_and_make_absolute("/proc/self/exe", &exedir) >= 0);
+
+        /* Check if we're running from the builddir. If so, use the compiled in path. */
+        if (path_startswith(exedir, ABS_BUILD_DIR))
+                return ABS_SRC_DIR "/test";
+
+        /* Try relative path, according to the install-test layout */
         assert_se(snprintf(testdir, sizeof(testdir), "%s/testdata", dirname(exedir)) > 0);
-        return testdir;
+        if (access(testdir, F_OK) >= 0)
+                return testdir;
+
+        fputs("ERROR: Cannot find testdata directory, set $SYSTEMD_TEST_DATA\n", stderr);
```

Comment:
We show this message for default directories only. But
```
SYSTEMD_TEST_DATA=/some/testdata ./libtool --mode=execute ./test-engine
...
Test1: (Trivial)
error: org.freedesktop.systemd1.NoSuchUnit: Unit c.service not found.
Assertion 'r == 0' failed at src/test/test-engine.c:61, function main(). Aborting.
Aborted (core dumped)
```
Maybe, we should call `access(testdir, F_OK)` in `TEST_DATA_DIR`

I mean something like:
```
d = getenv(SYSTEMD_TEST_DATA)
if !d
     d = get_exe_relative_testdata_dir()
if access(testdir, F_OK) >= 0
     return d
...something went wrong here...
```